enhancements/installer/storage-class-encrypted.md

```diff
@@ -0,0 +1,308 @@
+    // the format for a KMS key ARN is:
+    // arn:${Partition}:kms:${Region}:${Account}:key/${KeyId}
+    // +optional
+    KMSKeyARN string `json:"kmsKeyARN,omitempty"`
+}
+```
+#### Azure
+
+```go
+// AzureDiskEncryptionSet defines the configuration for a disk encryption set.
+type AzureDiskEncryptionSet struct {
+  // SubscriptionID defines the Azure subscription that contains the disk encryption set.
+  // When omitted, the subscription from the authenticating credentials of the CSI driver
+  // will be used.
+  // +optional 
+  SubscriptionID string `json:"subscriptionId"`
+  
+  // ResourceGroup defines the Azure resource group that contains the disk encryption set.
+  // When omitted, the cluster resource group from the cluster infrastructure object
+  // will be used.
+  // +optional
+  ResourceGroup string `json:"resourceGroup"`
+  
+  // Name is the name of the disk encryption set that will be set on the default storage class.
+  // +kubebuilder:validation:Required
+  // +kubebuilder:validation:MaxLength:=80
+  // +kubebuilder:validation:Pattern:=[a-zA-Z0-9_-]
+  // +required
+  Name string `json:"name"`
+}
+
+// AzureCSIDriverConfigSpec defines properties that can be configured for the Azure CSI driver.
+type AzureCSIDriverConfigSpec struct{
+  // DiskEncryptionSet sets the cluster default storage class to encrypt volumes with a
+  // customer-managed encryption set, rather than the default platform-managed keys.
+  // +optional
+  DiskEncryptionSet *AzureDiskEncryptionSet `json:"diskEncryptionSet,omitempty"`
+}
+
+```
+
+#### GCP
+
+```go
+// GCPKMSKeyReference gathers required fields for looking up a GCP KMS Key
+type GCPKMSKeyReference struct {
+  // Name is the name of the customer-managed encryption key to be used for disk encryption.
+  // +kubebuilder:validation:Required
+  // +kubebuilder:validation:Pattern:=[a-zA-Z0-9_-]{1,63}
+  Name string `json:"name"`
+  
+  // KeyRing is the name of the KMS Key Ring which the KMS Key belongs to.
+  // +kubebuilder:validation:Required
+  // +kubebuilder:validation:Pattern:=[a-zA-Z0-9_-]{1,63}
+  KeyRing string `json:"keyRing"`
+  
+  // ProjectID is the ID of the Project in which the KMS Key Ring exists.
+  // Defaults to the cluster project defined in the cluster infrastructure object, if not set.
+  // +optional
+  ProjectID string `json:"projectID,omitempty"`
+  
+  // Location is the GCP location in which the Key Ring exists.
+  // Defaults to global, if not set.
+  // +optional
+  Location string `json:"location,omitempty"`
+}
+
+// GCPCSIDriverConfigSpec defines properties that can be configured for the GCP CSI driver.
+type GCPCSIDriverConfigSpec struct {
+    // KMSKey sets the cluster default storage class to encrypt volumes with customer-supplied
+    // encryption keys, rather than the default keys managed by GCP. 
+    // +optional
+    KMSKey *GCPKMSKeyReference `json:"kmsKey,omitempty"`
```

Comment:
The machine API defines a [`KMSKeyServiceAccount`](https://github.com/openshift/api/blob/master/machine/v1beta1/types_gcpprovider.go#L192):

> KMSKeyServiceAccount is the service account being used for the
> encryption request for the given KMS key. If absent, the Compute
> Engine default service account is used.
> See https://cloud.google.com/compute/docs/access/service-accounts#compute_engine_service_account
> for details on the default service account.

Reviewing [the docs for the GCP driver,](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/6e3972c50121fd6a21f4aea85cabe59286898b51/README.md#createvolume-parameters) it does not appear to accept a service account, so I have omitted it from the API. It seems this parameter may be compute instance specific--in the storage case it would make sense to use the service account with which the csi driver is authenticating.

Reply:
I think this is alright, I don't see any way to pass in a separate service account to be used. The SC parameters just need the fields from GCPKMSKeyReference
https://cloud.google.com/kubernetes-engine/docs/how-to/using-cmek#storage-key